test/spec/Control/Biegunka/Source/GitSpec.hs

```diff
@@ -0,0 +1,148 @@
+              (repoRemote, repoLocal) <- buildRemoteRepo tmp
+              Git.askGit' tmp ["clone", repoRemote, repoLocal]
+              numberOfCommitsRemote <- numberOfCommits repoLocal
+              Git.askGit' repoLocal ["reset", "--hard", "HEAD~1"]
+              modify repoLocal
+              Git.updateGit repoRemote repoLocal Git.defaultGit
+              numberOfCommitsLocal <- numberOfCommits repoLocal
+              modifiedFiles repoLocal `shouldReturn` []
+              numberOfCommitsLocal `shouldBe` ((+) 1) `fmap` numberOfCommitsRemote
+
+          context "when failIfAhead flag is set" $
+            it "fails with exception" $ \tmp -> do
+              (repoRemote, repoLocal) <- buildRemoteRepo tmp
+              Git.askGit' tmp ["clone", repoRemote, repoLocal]
+              Git.askGit' repoLocal ["reset", "--hard", "HEAD~1"]
+              modify repoLocal
+              Git.updateGit repoRemote repoLocal (Git.defaultGit { Git._failIfAhead = True })
+                `shouldThrow` sourceException
+
+      context "when repo has a dirty state" $
+        it "fails with exception" $ \tmp -> do
+          (repoRemote, repoLocal) <- buildRemoteRepo tmp
+          Git.askGit' tmp ["clone", repoRemote, repoLocal]
+          Git.askGit' repoLocal ["reset", "--soft", "HEAD~1"]
+          Git.updateGit repoRemote repoLocal (Git.defaultGit { Git._failIfAhead = True })
+            `shouldThrow` sourceException
+
+
+withBiegunkaDirectory :: (FilePath -> IO a) -> IO a
+withBiegunkaDirectory action = do
+ str <- liftM (take 10 . randomRs ('a','z')) newStdGen
+ withSystemTempDirectory ("biegunka-" ++ str ++ "-") action
+
+currentBranch :: FilePath -> IO (Maybe String)
+currentBranch path = (listToMaybe . lines) `fmap` Git.askGit path ["rev-parse", "--abbrev-ref", "HEAD"]
+
+modifiedFiles :: FilePath -> IO [String]
+modifiedFiles path = lines `fmap` Git.askGit path ["diff-index", "HEAD"]
+
+buildRemoteRepo :: FilePath -> IO (FilePath, FilePath)
+buildRemoteRepo path = do
+  let repo = path </> "test-repo-remote"
+  let file = repo </> "file"
+  createDirectory repo
+  Git.askGit' repo ["init"]
+  setCredentials repo
+  Git.askGit' repo ["commit", "--allow-empty", "--message", "initial"]
+  writeFile file "hi\nthere\n!\n"
+  Git.askGit' repo ["add", file]
+  Git.askGit' repo ["commit", "--message", "Change 1"]
+  appendFile file "never\nmind\n!\n"
+  Git.askGit' repo ["commit", "-a", "--message", "Change 2"]
+  appendFile file "just\nkidding\n!\n"
+  Git.askGit' repo ["commit", "-a", "--message", "Change 3"]
+  return (repo, path </> "test-repo-local")
+
+sourceException :: Selector SourceException
+sourceException = const True
+
+prependFile :: FilePath -> String -> IO ()
+prependFile path s = do
+  content <- readFile path
+  length content `seq` writeFile path $ s ++ content
+
+modify :: FilePath -> IO ()
+modify path = do
+  setCredentials path
+  prependFile (path </> "file") "new string"
```

Comment:
This is highly non-obvious. That detail should either be documented or sidestepped by changing the way to make local modifications to e.g. writing to another file.